bemenu: init at 0.1.0

thiagokokada opened the pull request:
###### Motivation for this change

Bemenu is a dynamic menu library and client program inspired by dmenu with support for wayland compositors (and also X11 and ncurses).

Build and tested in NixOS `19.03.172392.6d7ed96429` in X11 (`i3wm`), wayland (`sway`) and ncurses (`kitty`).

There seems to be a package for [bemenu already in master](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/window-managers/orbment/bemenu.nix), however this is based in a very old commit (~2017) when `bemenu` used to only run in `orbment`. Nowadays `bemenu` also runs in other wayland compositors like `sway`, so it is a interesting alternative to `dmenu`, `rofi` for wayland users.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

alexarice commented:
Looks really interesting as I have been having problems with dmenu in wayland.
Probably better to use `pname` instead of `name`

alexarice commented:
It seems there are options upstream to build for different renderers. Is this building all 3?

thiagokokada commented:
@alexarice Yeah, it is building for all 3 renderers. You can change between them using `BEMENU_BACKEND`.

alexarice commented:
Ah I was confused as the readme has changed since 0.1.0

thiagokokada commented:
@alexarice Whats the difference between `pname` and `name`?

alexarice commented:
If you write
```
pname = "bemenu";
version = "0.1.0";
```
then it automatically sets `name = ${pname}-${version}`
It doesn't really change anything, it's just a bit neater

thiagokokada commented:
@alexarice Cool, done!

etu commented on maintainers/maintainer-list.nix:
```diff
@@ -5540,4 +5540,9 @@
     github = "tesq0";
     name = "Mikolaj Galkowski";
 	};
+  thiagokokada = {
```
It would be nice if you added the maintainership as a separate commit so they are independent of each other.

etu replied:
Also, this list is supposed to be in alphabetical order, so move your entry upwards. I can see that some haven't followed this, but fix your entry at least :-)

alexarice replied:
Does the same apply to all-packages.nix?

etu replied:
That one is harder to keep sorted since it's huge, but at least there are attempts. I usually do a dirlist and place it close to the package before or after in the dirlist of where I put it. In this case close to `bb` or `bibletime` or something like that...

alexarice commented:
Both executables seem to work to me. It would be nice but not essential if docs worked and there were options to enable/disable backends

thiagokokada committed:
maintainers: Add thiagokokada to maintainer-list

thiagokokada committed:
bemenu: init at 0.1.0

Bemenu is a dynamic menu library and client program inspired by dmenu
with support for wayland compositors (and also X11 and ncurses).

Build and tested in NixOS 19.03.172392.6d7ed96429 in X11 (i3wm), wayland
(sway) and ncurses (kitty).

thiagokokada commented:
@etu Made the changes.

@alexarice I added support to build only the backends the user wants. I tried to add documentation support, however the only output I got was either LaTeX or html, so I don't think it is much useful.

alexarice commented on pkgs/applications/misc/bemenu/default.nix:
```diff
@@ -0,0 +1,45 @@
+{ stdenv, fetchFromGitHub, cairo, cmake, libxkbcommon
+, pango, fribidi, harfbuzz, pcre, pkgconfig
+, ncursesSupport ? true, ncurses ? null
```
Is this necessary since nix is lazy?

thiagokokada replied:
I was thinking about this too, however is there some good way to declare x11 support since it depends on two different group of packages (`x11` and `xorg`)?

The rest I just put to maintain the same way.

alexarice replied:
Checked this and don't see the need for the `? null` though I could be wrong

alexarice replied:
my first comment was aimed at the `? null` bit and not the `ncursesSupport` bit if that wasn't clear as I'm not sure what you mean

alexarice commented on pkgs/applications/misc/bemenu/default.nix:
```diff
@@ -0,0 +1,45 @@
+    ++ optionals waylandSupport [ wayland ]
+    ++ optionals x11Support [
+      xlibs.libX11 xlibs.libXinerama xlibs.libXft
+      xorg.libXdmcp xorg.libpthreadstubs xorg.libxcb
```
Seems to build fine without the xorg dependencies for me though someone else should probably check this as I'm not actually running X (it does claim to be building the x11 backend though)

thiagokokada replied:
@alexarice Sure, this makes sense. Removed the null from dependencies.

alexarice replied:
Does your build break without the `xorg` dependencies, or are they needed at run time or something similar

thiagokokada replied:
Setting `x11Support` to false correctly builds the binary without X11 support.

thiagokokada replied:
Well, it really breaks eval when removing those null checks. I re-added them, I think the tests should pass now.

alexarice replied:
I'm still not being clear, I can remove the dependencies starting with xorg from the nix file and build it with x11 support

etu commented:
@thiagokokada The commit message should also be: `bemenu: init at 0.1.0`

etu commented:
Also, something in this PR seems to break eval. But there's no details link.

So here goes @GrahamcOfBorg eval

thiagokokada commented:
@etu Fixed commit message. However, I have no idea why eval is breaking. Where could I see output from @GrahamcOfBorg?

thiagokokada commented:
@GrahamcOfBorg

thiagokokada commented:
@alexarice @etu This should build now (at least seems to build fine locally).

etu commented:
The eval is happy now at least :-)

@GrahamcOfBorg build bemenu

thiagokokada commented:
@alexaice Yeah, I know you can build with less dependencies, however you
get some warnings during build for missing libraries so I decided to be in
the safe side.

On Fri, May 3, 2019, 04:45 alexarice <notifications@github.com> wrote:

> *@alexarice* commented on this pull request.
> ------------------------------
>
> In pkgs/applications/misc/bemenu/default.nix
> <https://github.com/NixOS/nixpkgs/pull/60661#discussion_r280685490>:
>
> > +    sha256 = "11h55m9dx6ai12pqij52ydjm36dvrcc856pa834njihrp626pl4w";
> +  };
> +
> +  nativeBuildInputs = [ cmake pkgconfig pcre ];
> +
> +  buildInputs = with stdenv.lib; [
> +    cairo
> +    fribidi
> +    harfbuzz
> +    libxkbcommon
> +    pango
> +  ] ++ optionals ncursesSupport [ ncurses ]
> +    ++ optionals waylandSupport [ wayland ]
> +    ++ optionals x11Support [
> +      xlibs.libX11 xlibs.libXinerama xlibs.libXft
> +      xorg.libXdmcp xorg.libpthreadstubs xorg.libxcb
>
> I'm still not being clear, I can remove the dependencies starting with
> xorg from the nix file and build it with x11 support
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/60661#discussion_r280685490>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGOEN5IPJLQWTHRC63C4YDPTPUQFANCNFSM4HJVXEGA>
> .
>

alexarice commented:
Fair enough, It doesn't matter anyway I guess

thiagokokada commented:
@etu @alexarice Can I get this PR merged? Or is there some other issue that needs to be fixed?

alexarice commented:
@thiagokokada I am not able to merge things but don't see any problems

etu commented:
@thiagokokada I've not been able to try this program locally. It builds fine but when I run it. Nothing happens. I've tried to supply the env-variable to force it to run in ncurses mode. And just no output. So I'm not sure how it's supposed to be used...

alexarice commented:
@etu have you tried `bemenu-run`? both that and `j4-dmenu-desktop --dmenu="result/bin/bemenu"` work for me on sway

etu commented:
@alexarice Thanks, didn't notice that binary. I tabcompleted to bemenu and got confused by it's behaviour. But it all makes sense now :-)

NilsIrl commented:
Why does it still not appear on the package search?

https://nixos.org/nixos/packages.html?channel=nixpkgs-unstable&query=bemenu

NilsIrl commented:
Also why is it that when I test it locally from nixpkgs there is not font (all characters are unknown sign) but when I install it with `nix-env -iA` it works?
